Event Driven Architecture (#407)
developer-overheid-nl/don-site: blog/draft/event-driven.md

---
authors: [martin-van-der-plas,]
tags: [api, eda, fds, webhooks, cloudevents]
description: Steeds meer overheidsorganisaties stappen over naar event-driven architecturen (EDA). Waarom? Omdat het beter schaalbaar is, sneller reageert op veranderingen en beter past bij een moderne digitale overheid waarin systemen onderling proactief samenwerken op basis van gebeurtenissen, in plaats van reactieve vraag-antwoordstructuren. In dit artikel duiken we in de kern van EDA binnen de Nederlandse overheid en bespreken we Stelsels, Standaarden en Voorzieningen als CloudEvents, DigiLevering, DigiMelding, en Webhooks zoals die terugkomen in de NL API Strategie.
---
import { Blockquote } from "@rijkshuisstijl-community/components-react";

> # De toekomstige van de Overheid is Event Driven

Reviewer: ```suggestion
# De toekomst van de Overheid is Event Driven
```

![Schematisch beeld van de Event Driven Overheid](./img/event-driven-overheid.png)
*Schematisch beeld van de Event Driven Overheid*

> Steeds meer overheidsorganisaties stappen over naar [event-driven architecturen (EDA)](../kennisbank/apis/architectuur/eda). Waarom? Omdat het beter schaalbaar is, sneller reageert op veranderingen en beter past bij een moderne digitale overheid waarin systemen onderling proactief samenwerken op basis van gebeurtenissen, in plaats van reactieve vraag-antwoordstructuren. In dit artikel duiken we in de kern van EDA binnen de Nederlandse overheid en bespreken we Stelsels, Standaarden en Voorzieningen als [CloudEvents](../kennisbank/apis/architectuur/cloudevents), [Webhooks](../kennisbank/apis/architectuur/cloudevents) DigiLevering en DigiMelding.

Reviewer: ```suggestion
Steeds meer overheidsorganisaties stappen over naar [eventgedreven architecturen (EDA)](../kennisbank/apis/architectuur/eda). Waarom? Omdat het beter schaalbaar is, sneller reageert op veranderingen en beter past bij een moderne digitale overheid waarin systemen onderling proactief samenwerken op basis van gebeurtenissen, in plaats van reactieve vraag-antwoordstructuren. In dit artikel duiken we in de kern van EDA binnen de Nederlandse overheid en bespreken we Stelsels, Standaarden en Voorzieningen als [CloudEvents](../kennisbank/apis/architectuur/cloudevents), [Webhooks](../kennisbank/apis/architectuur/cloudevents) DigiLevering en DigiMelding.
```

## Context - Huidige situatie

<Blockquote
  variation="pink-background"
  attribution="- Forum Standaardisatie -"
>
Op 25 september 2025 heeft het Forum Standaardisatie ingestemd met het advies aan het OBDO om het *NL GOV profile for CloudEvents* verplicht te stellen (“Pas toe of leg uit”). Daarnaast is het predicaat **‘Uitstekend beheer’** toegekend voor deze standaard.
</Blockquote>

### **CloudEvents: gestandaardiseerd eventformaat**

Een belangrijke ontwikkeling is de standaardisering van eventformaat via [**CloudEvents**](https://cloudevents.io/). CloudEvents is een CNCF-standaard die beschrijft hoe events uniform verpakt kunnen worden, onafhankelijk van bron of transportprotocol. CloudEvents specificeert onder andere:

- De **bron** van het event
- Het **type** event (bijvoorbeeld nl.basisregistratie.persoon.geboorte)

> - De **tijdstip** en **ID**

Reviewer: ```suggestion
- Het **tijdstip** en **ID**
```

- bij de BRK notificaties bewust gekozen voor een pull mechanisme met een api. zie ook https://www.kadaster.nl/zakelijk/registraties/basisregistraties/brk/brk-notificaties

Verder zijn de best practices erg afhankelijk van de use case die je voor ogen hebt. hierin zou je moeten adresseren:

- is het echt realtime of mag er een vertraging zijn en wat is dan acceptabel?
- wie bepaald het tempo van verwerking van events (provider of client)
- is er wel of geen garantie van aflevering nodig?
- hoe vind interne verwerking van events vervolgens plaats bij provider en of client?
- welke eisen zijn er aan interne verwerking en standaarden en patronen zoals JMS , Kafka , mq amqp

> Tip voor developers: experimenteer eens met CloudEvents in combinatie met een [local webhook listener zoals webhook.site](https://webhook.site/), of gebruik [ngrok](https://ngrok.com/) om een publieke endpoint te creëren voor test-events uit overheidsdomeinen.

Wil je sparren over hoe jouw organisatie events kan inzetten? Of heb je een vraag over implementatie in je API? Deel je ervaringen op [developer.overheid.nl/community](https://developer.overheid.nl/community).

### Soorten Events

> Er is een belangrijk onderscheid in soorten events:

Reviewer: Is het niet beter om gebeurtenissen hier te gebruiken?

#### Events als reactie op een bericht (request based event)

@@@

- Er is geen vastgestelde status van een stelsel. Op elk moment kunnen partijen besluiten om zelf mee te gaan doen. (dynamic registration)
- De initiele message heeft alle metadata in zich om verdere notificaties af te handelen zoals shared secrets (of asymetrische versleuteling) en callback uri's
- Notificaties worden op basis van de initiele message naar 1 of meerdere eindpoints gedeeld. (bij voorkeur op basis van webhooks - wellicht moet mail ook kunnen)
- Notificaties zijn in deze opzet altijd een afgeleide van een message. De aanbieder van de service (waard de message wordt afgeleverd) biedt de notificatie functie als aanvullende dienst aan. Merk op dat deze aanvullende notificatie functie gelijk kan zijn aan de notificatie functie binnen het stelsel.

Bijvoorbeeld:

- Een gebruiker stuurt een aanvraag → het systeem stuurt een “aanvraag ontvangen”-event.
- Een notificatie over de status van verwerking (“verzonden”, “afgewezen”)

Dit is een soort *event-after-request*: de gebeurtenis is een gevolg van een actieve handeling.

#### Events als registratiegebeurtenis (object based event)

> @@@ Stelsel based Notificeren:

Reviewer: ```suggestion
@@@ Stelsel-based Notificeren:
```

of gewoon

```suggestion
@@@ Nnotificeren op stelselbasis:
```

- Gebeurtenis + Identificatie: Bij zo'n 'informatiearme notificatie' worden minimaal gebeurtenisgegevens verstrekt. Met behulp van verstrekte identificerende gegevens kan een consumer eventueel aanvullende gegevens gaan opvragen.

### Webhooks: Real-time event notificaties  

**Webhooks** zijn een van de meest gebruikte methoden om EDA te implementeren. Ze maken **real-time notificaties** via **HTTP** mogelijk. In plaats van constant een API te pollen op updates, ontvangen consumenten automatisch een notificatie zodra een event plaatsvindt.  

**Waarom Webhooks gebruiken?**  

- **Vermindert API-verzoeken** – Geen constante polling nodig.  
- **Real-time updates** – Events worden direct doorgegeven.  
- **Lichtgewicht** – Eenvoudig te implementeren met HTTP POST-verzoeken.  

> *Meer over Webhooks:*  - [Webhooks uitgelegd (GitHub)](https://docs.github.com/en/webhooks)  

### OAuth in Event-Driven Architectures  

> Beveiliging is een cruciaal aspect in EDA, vooral bij het **exposeren van event-driven API’s**. **OAuth 2.0** en **OpenID Connect (OIDC)** zijn de meest gebruikte standaarden voor **authenticatie en autorisatie** in event-gedreven interacties.  

Reviewer: Ik vind het prima zo (biased 😄) maar misschien ook gewoon ruimte laten voor API-keys (staat soort van in de webhooks guidelines). En mTLS misschien ook noemen?

- Een **API-ecosysteem met embedded webhooks in OpenAPI**

Om dat goed te laten werken, zijn afspraken nodig over:

- **Standaardisatie** (zoals CloudEvents profielen voor overheden)
- **Beveiliging** van webhooks (OAuth, retry-strategieën, verificatie)
- **Beheer** van abonnementen

**Event-Driven Architecture** verandert moderne API-ecosystemen door **schaalbare, real-time en losgekoppelde** interacties mogelijk te maken. Door gebruik te maken van **CloudEvents, Webhooks en OAuth** kunnen organisaties robuuste event-gedreven oplossingen ontwikkelen die naadloos aansluiten op de **Nederlandse API Strategie**.  

### Uitdagingen @@@ is dit wel handig

@@@ herschrijven van dit stuk
Standaarden en afspraken over inderdaad het Abonneren, maar ook over bv.:

- Het concept Gebeurtenis; wat is nu een gebeurtenis, welke richtlijnen hanteren om die per bron te identificeren/benoemen?
- Hoe omgaan met privacy-gevoelige gebeurtenissen als adoptie of geslachtswijziging; deze te maskeren via niet herleidbare aliassen?
- Is het altijd Gegevensarm notificeren of zijn er usecases waarbij je niet ontkomt aan Gegevensrijk notificeren? Zoja, hoe dan omgaan met gegevensautorisatie per afnemer?

> - Hoe omgaan met filtering van Notificaties naar doelgroep; bv. als een afnemer alleen notificaties mag ontvangen van personen in een bepaalde leefdtijdsgroep?

Reviewer: ```suggestion
- Hoe omgaan met filtering van Notificaties naar doelgroep; bv. als een afnemer alleen notificaties mag ontvangen van personen in een bepaalde leeftijdsgroep?
```
